File: contractsdk/cpp/example/erc20.cc
#include "xchain/xchain.h"

struct ERC20 : public xchain::Contract {};

const std::string BALANCEPRE = "balanceOf_";
const std::string ALLOWANCEPRE = "allowanceOf_";
const std::string MASTERPRE = "owner";

DEFINE_METHOD(ERC20, initialize) {
    xchain::Context* ctx = self.context();
    const std::string& caller = ctx->initiator();
    if (caller.empty()) {
        ctx->error("missing caller");
        return;
    }
    const std::string& totalSupply = ctx->arg("totalSupply");
    if (totalSupply.empty()) {
        ctx->error("missing totalSupply");
        return;
    }

    std::string key = BALANCEPRE + caller;
    ctx->put_object("totalSupply", totalSupply);
    ctx->put_object(key, totalSupply);

    std::string master = MASTERPRE;
    ctx->put_object(master, caller);
    ctx->ok("initialize success");
}

DEFINE_METHOD(ERC20, mint) {
    xchain::Context* ctx = self.context();
    const std::string& caller = ctx->initiator();
    if (caller.empty()) {
        ctx->error("missing caller");
        return;
    }

    std::string master;
    if (!ctx->get_object(MASTERPRE, &master)) {
        ctx->error("missing master");
        return;
    }
    if (master != caller) {
        ctx->error("only the person who created the contract can mint");
        return;
    }

    const std::string& increaseSupply = ctx->arg("amount");
    if (increaseSupply.empty()) {
        ctx->error("missing amount");
        return;
    }

    std::string value;
    if (!ctx->get_object("totalSupply", &value)) {
        ctx->error("get totalSupply error");
        return;
    }
    
    int increaseSupplyint = atoi(increaseSupply.c_str());
    int valueint = atoi(value.c_str());
    int totalSupplyint = increaseSupplyint + valueint;
    char buf[32];
    snprintf(buf, 32, "%d", totalSupplyint);
    ctx->put_object("totalSupply", buf); 
    
    std::string key = BALANCEPRE + caller;
    if (!ctx->get_object(key, &value)) {
        ctx->error("get caller balance error");
        return;
    }
    valueint = atoi(value.c_str());
    int callerint = increaseSupplyint + valueint;
    snprintf(buf, 32, "%d", callerint);
    ctx->put_object(key, buf); 
    
    ctx->ok(buf);
}

DEFINE_METHOD(ERC20, totalSupply) {
    xchain::Context* ctx = self.context();
    std::string value;
    if (ctx->get_object("totalSupply", &value)) {
        ctx->ok(value);
    } else {
        ctx->error("key not found");
    }
}

DEFINE_METHOD(ERC20, balance) {
    xchain::Context* ctx = self.context();
    const std::string& caller = ctx->arg("caller");
    if (caller.empty()) {
        ctx->error("missing caller");
        return;
    }
    
    std::string key = BALANCEPRE + caller;
    std::string value;
    if (ctx->get_object(key, &value)) {
        ctx->ok(value);
    } else {
        ctx->error("key not found");
    }
}

DEFINE_METHOD(ERC20, allowance) {
    xchain::Context* ctx = self.context();
    const std::string& from = ctx->arg("from");
    if (from.empty()) {
        ctx->error("missing from");
        return;
    }
   
    const std::string& to = ctx->arg("to");
    if (to.empty()) {
        ctx->error("missing to");
        return;
    }

    std::string key = ALLOWANCEPRE + from + "_" + to;
    std::string value;
    if (ctx->get_object(key, &value)) {
        ctx->ok(value);
    } else {
        ctx->error("key not found");
    }
}

DEFINE_METHOD(ERC20, transfer) {
    xchain::Context* ctx = self.context();
    const std::string& from = ctx->arg("from");
    if (from.empty()) {
        ctx->error("missing from");
        return;
    }
   
    const std::string& to = ctx->arg("to");
    if (to.empty()) {
        ctx->error("missing to");
        return;
    }

    const std::string& token_str = ctx->arg("token");
    if (token_str.empty()) {
        ctx->error("missing token");
        return;
    }
    int token = atoi(token_str.c_str());

    std::string from_key = BALANCEPRE + from;
    std::string value;
    int from_balance = 0;
    if (ctx->get_object(from_key, &value)) {
        from_balance = atoi(value.c_str()); 
        if (from_balance < token) {
            ctx->error("The balance of from not enough");
            return;
        }  
    } else {
        ctx->error("key not found");
        return;
    }

    std::string to_key = BALANCEPRE + to;
    int to_balance = 0;
    if (ctx->get_object(to_key, &value)) {
        to_balance = atoi(value.c_str());
    }
   
    from_balance = from_balance - token;
    to_balance = to_balance + token;
   
    char buf[32]; 
    snprintf(buf, 32, "%d", from_balance);
    ctx->put_object(from_key, buf);
    snprintf(buf, 32, "%d", to_balance);
    ctx->put_object(to_key, buf);

    ctx->ok("transfer success");
}

DEFINE_METHOD(ERC20, transferFrom) {
    xchain::Context* ctx = self.context();
    const std::string& from = ctx->arg("from");
    if (from.empty()) {
        ctx->error("missing from");
        return;
    }
  
    const std::string& caller = ctx->arg("caller");
    if (caller.empty()) {
        ctx->error("missing caller");
        return;
    }

    const std::string& to = ctx->arg("to");
    if (to.empty()) {
        ctx->error("missing to");
        return;
    }

    const std::string& token_str = ctx->arg("token");
    if (token_str.empty()) {
        ctx->error("missing token");
        return;
    }
    int token = atoi(token_str.c_str());

    std::string allowance_key = ALLOWANCEPRE + from + "_" + caller;
    std::string value;
    int allowance_balance = 0;
    if (ctx->get_object(allowance_key, &value)) {
        allowance_balance = atoi(value.c_str()); 
        if (allowance_balance < token) {
            ctx->error("The allowance of from_to not enough");
            return;
        }  
    } else {
        ctx->error("You need to add allowance from_to");
        return;
    }

    std::string from_key = BALANCEPRE + from;
    int from_balance = 0;
    if (ctx->get_object(from_key, &value)) {
        from_balance = atoi(value.c_str()); 
        if (from_balance < token) {
            ctx->error("The balance of from not enough");
            return;
        }  
    } else {
        ctx->error("From no balance");
        return;
    }

    std::string to_key = BALANCEPRE + to;
    int to_balance = 0;
    if (ctx->get_object(to_key, &value)) {
        to_balance = atoi(value.c_str());
    }
   
    from_balance = from_balance - token;
    to_balance = to_balance + token;
    allowance_balance = allowance_balance - token;

    char buf[32]; 
    snprintf(buf, 32, "%d", from_balance);
    ctx->put_object(from_key, buf);
    snprintf(buf, 32, "%d", to_balance);
    ctx->put_object(to_key, buf);
    snprintf(buf, 32, "%d", allowance_balance);
    ctx->put_object(allowance_key, buf);

    ctx->ok("transferFrom success");
}

DEFINE_METHOD(ERC20, approve) {
    xchain::Context* ctx = self.context();
    const std::string& from = ctx->arg("from");
    if (from.empty()) {
        ctx->error("missing from");
        return;
    }
   
    const std::string& to = ctx->arg("to");
    if (to.empty()) {
        ctx->error("missing to");
        return;
    }

    const std::string& token_str = ctx->arg("token");
    if (token_str.empty()) {
        ctx->error("missing token");
        return;
    }
    int token = atoi(token_str.c_str());

    std::string from_key = BALANCEPRE + from;
    std::string value;
    if (ctx->get_object(from_key, &value)) {
        int from_balance = atoi(value.c_str()); 
        if (from_balance < token) {
            ctx->error("The balance of from not enough");
            return;
        }  
    } else {
        ctx->error("From no balance");
        return;
    }

    std::string allowance_key = ALLOWANCEPRE + from + "_" + to;
    int allowance_balance = 0;
    if (ctx->get_object(allowance_key, &value)) {
        allowance_balance = atoi(value.c_str()); 
    }

    allowance_balance = allowance_balance + token;
   
    char buf[32]; 
    snprintf(buf, 32, "%d", allowance_balance);
    ctx->put_object(allowance_key, buf);

    ctx->ok("approve success");
}

File: core/contractsdk/java/example/erc20/src/main/java/com/baidu/xuper/Erc20.java
package com.baidu.xuper;

import java.math.BigInteger;
import java.util.Arrays;

import com.baidu.xuper.Context;
import com.baidu.xuper.Contract;
import com.baidu.xuper.ContractMethod;
import com.baidu.xuper.Driver;
import com.baidu.xuper.Response;

/**
 * Erc20
 *
 */
public class Erc20 implements Contract
{
    static final String BALANCEPRE = "balanceOf_";
    static final String ALLOWANCEPRE = "allowanceOf_";
    static final String MASTERPRE = "owner";
    static final String TOTALSUPPLY = "totalSupply";

    @Override
    @ContractMethod
    public Response initialize(Context ctx) {
        String caller = ctx.caller();
        if (caller.isEmpty()){
            return Response.error("missing caller");
        }

        byte[] totalSupplyByte = ctx.args().get("totalSupply");
        if (totalSupplyByte == null) {
            return Response.error("missing totalSupply");
        }
        String totalSupplyStr = new String(totalSupplyByte);
        BigInteger totalSupply = new BigInteger(totalSupplyStr);
        if (totalSupply.signum() == -1) {
            return Response.error("totalSupply must not be negative");
        }

        String balanceKey = BALANCEPRE + caller;

        ctx.putObject(TOTALSUPPLY.getBytes(), totalSupply.toByteArray());
        ctx.putObject(balanceKey.getBytes(), totalSupply.toByteArray());
        ctx.putObject(MASTERPRE.getBytes(), caller.getBytes());

        return Response.ok("initialize success".getBytes());
    }

    @ContractMethod
    public Response mint(Context ctx) {
        String caller = ctx.caller();
        if (caller.isEmpty()){
            return Response.error("missing caller");
        }

        byte[] ownerByte = ctx.getObject(MASTERPRE.getBytes());
        if (ownerByte == null) {
            return Response.error("no owner found");
        }

        if (!Arrays.equals(caller.getBytes(), ownerByte)){
            return Response.error("only the person who created the contract can mint");
        }

        byte[] increaseSupplyByte = ctx.args().get("amount");
        if (increaseSupplyByte == null) {
            return Response.error("missing increaseSupply");
        }
        String increaseSupplyStr = new String(increaseSupplyByte);
        BigInteger increaseSupply = new BigInteger(increaseSupplyStr);
        if (increaseSupply.signum() == -1) {
            return Response.error("amount must not be negative");
        }

        byte[] totalSupplyByte = ctx.getObject(TOTALSUPPLY.getBytes());
        if (totalSupplyByte == null) {
            return Response.error("no totalSupply found");
        }
        BigInteger totalSupply = new BigInteger(totalSupplyByte);
        BigInteger totalSupplyNow = totalSupply.add(increaseSupply);

        ctx.putObject(TOTALSUPPLY.getBytes(), totalSupplyNow.toByteArray());

        String balanceKey = BALANCEPRE + caller;
        byte[] callerBalanceByte = ctx.getObject(balanceKey.getBytes());
        if (callerBalanceByte == null) {
            return Response.error("no caller found");
        }
        BigInteger callerBalance = new BigInteger(callerBalanceByte);
        BigInteger callerBalanceNow = callerBalance.add(increaseSupply);

        ctx.putObject(balanceKey.getBytes(), callerBalanceNow.toByteArray());

        return Response.ok("mint success".getBytes());
    }

    @ContractMethod
    public Response totalSupply(Context ctx) {
        byte[] value = ctx.getObject(TOTALSUPPLY.getBytes());
        if (value == null) {
            return Response.error("key TOTALSUPPLY not found)");
        }
        BigInteger totalSupply = new BigInteger(value);

        return Response.ok(totalSupply.toString().getBytes());
    }

    @ContractMethod
    public Response balance(Context ctx) {
        byte[] accountByte = ctx.args().get("account");
        if (accountByte == null) {
            return Response.error("missing account");
        }
        String account = new String(accountByte);

        String balanceKey = BALANCEPRE + account;
        byte[] value = ctx.getObject(balanceKey.getBytes());
        if (value == null) {
            return Response.error("key " + account + " not found");
        }
        BigInteger balance = new BigInteger(value);

        return Response.ok(balance.toString().getBytes());
    }

    @ContractMethod
    public Response allowance(Context ctx) {
        byte[] fromByte = ctx.args().get("from");
        if (fromByte == null) {
            return Response.error("missing from");
        }
        String from = new String(fromByte);

        byte[] toByte = ctx.args().get("to");
        if (toByte == null) {
            return Response.error("missing to");
        }
        String to = new String(toByte);

        String allowanceKey = ALLOWANCEPRE + from + "_" + to;
        byte[] value = ctx.getObject(allowanceKey.getBytes());
        if (value == null) {
            return Response.error("key " + allowanceKey + " not found");
        }
        BigInteger allowance = new BigInteger(value);

        return Response.ok(allowance.toString().getBytes());
    }

    @ContractMethod
    public Response owner(Context ctx) {
        byte[] ownerByte = ctx.getObject(MASTERPRE.getBytes());
        if (ownerByte == null) {
            return Response.error("no owner found");
        }

        return Response.ok(ownerByte);
    }

    @ContractMethod
    public Response transfer(Context ctx) {
        String from = ctx.caller();
        if (from.isEmpty()){
            return Response.error("missing from");
        }

        byte[] toByte = ctx.args().get("to");
        if (toByte == null) {
            return Response.error("missing to");
        }
        String to = new String(toByte);

        byte[] tokenByte = ctx.args().get("token");
        if (tokenByte == null) {
            return Response.error("missing token");
        }
        String tokenAmountStr = new String(tokenByte);
        BigInteger tokenAmount = new BigInteger(tokenAmountStr);
        if (tokenAmount.signum() == -1) {
            return Response.error("token must not be negative");
        }

        String fromKey = BALANCEPRE + from;
        byte[] fromBalanceByte = ctx.getObject(fromKey.getBytes());
        if (fromBalanceByte == null) {
            return Response.error("no from found");
        }
        BigInteger fromBalance = new BigInteger(fromBalanceByte);
        if (fromBalance.compareTo(tokenAmount) == -1){
            return Response.error("no enough balance");
        }

        String toKey = BALANCEPRE + to;
        byte[] toBalanceByte = ctx.getObject(toKey.getBytes());
        BigInteger toBalance;
        if (toBalanceByte == null) {
            toBalance = BigInteger.valueOf(0);
        } else {
            toBalance = new BigInteger(toBalanceByte);
        }

        BigInteger fromBalanceNow = fromBalance.subtract(tokenAmount);
        BigInteger toBalanceNow = toBalance.add(tokenAmount);

        ctx.putObject(fromKey.getBytes(), fromBalanceNow.toByteArray());
        ctx.putObject(toKey.getBytes(), toBalanceNow.toByteArray());

        return Response.ok("transfer success".getBytes());
    }

    @ContractMethod
    public Response transferFrom(Context ctx) {
        String caller = ctx.caller();
        if (caller.isEmpty()){
            return Response.error("missing caller");
        }

        byte[] fromByte = ctx.args().get("from");
        if (fromByte == null) {
            return Response.error("missing from");
        }
        String from = new String(fromByte);

        byte[] toByte = ctx.args().get("to");
        if (toByte == null) {
            return Response.error("missing to");
        }
        String to = new String(toByte);

        byte[] tokenByte = ctx.args().get("token");
        if (tokenByte == null) {
            return Response.error("missing token");
        }
        String tokenAmountStr = new String(tokenByte);
        BigInteger tokenAmount = new BigInteger(tokenAmountStr);
        if (tokenAmount.signum() == -1) {
            return Response.error("token must not be negative");
        }

        String allowanceKey = ALLOWANCEPRE + from + "_" + caller;
        byte[] value = ctx.getObject(allowanceKey.getBytes());
        if (value == null) {
            return Response.error("key " + allowanceKey + " not found");
        }
        BigInteger allowance = new BigInteger(value);
        if (allowance.compareTo(tokenAmount) == -1){
            return Response.error("The allowance is not enough");
        }

        String fromKey = BALANCEPRE + from;
        byte[] fromBalanceByte = ctx.getObject(fromKey.getBytes());
        if (fromBalanceByte == null) {
            return Response.error("no from found");
        }
        BigInteger fromBalance = new BigInteger(fromBalanceByte);
        if (fromBalance.compareTo(tokenAmount) == -1){
            return Response.error("The balance of from is not enough");
        }

        String toKey = BALANCEPRE + to;
        byte[] toBalanceByte = ctx.getObject(toKey.getBytes());
        BigInteger toBalance;
        if (toBalanceByte == null) {
            toBalance = BigInteger.valueOf(0);
        } else {
            toBalance = new BigInteger(toBalanceByte);
        }

        BigInteger fromBalanceNow = fromBalance.subtract(tokenAmount);
        BigInteger toBalanceNow = toBalance.add(tokenAmount);
        BigInteger allowanceBalanceNow = allowance.subtract(tokenAmount);

        ctx.putObject(fromKey.getBytes(), fromBalanceNow.toByteArray());
        ctx.putObject(toKey.getBytes(), toBalanceNow.toByteArray());
        ctx.putObject(allowanceKey.getBytes(), allowanceBalanceNow.toByteArray());

        return Response.ok("transferFrom success".getBytes());
    }

    @ContractMethod
    public Response approve(Context ctx) {
        String from = ctx.caller();
        if (from.isEmpty()){
            return Response.error("missing caller");
        }

        byte[] toByte = ctx.args().get("to");
        if (toByte == null) {
            return Response.error("missing to");
        }
        String to = new String(toByte);

        byte[] tokenByte = ctx.args().get("token");
        if (tokenByte == null) {
            return Response.error("missing token");
        }
        String tokenAmountStr = new String(tokenByte);
        BigInteger tokenAmount = new BigInteger(tokenAmountStr);
        if (tokenAmount.signum() == -1) {
            return Response.error("token must not be negative");
        }

        String fromKey = BALANCEPRE + from;
        byte[] fromBalanceByte = ctx.getObject(fromKey.getBytes());
        if (fromBalanceByte == null) {
            return Response.error("no from found");
        }
        BigInteger fromBalance = new BigInteger(fromBalanceByte);
        if (fromBalance.compareTo(tokenAmount) == -1){
            return Response.error("The balance of from not enough");
        }

        String allowanceKey = ALLOWANCEPRE + from + "_" + to;
        byte[] allowanceBalanceByte = ctx.getObject(allowanceKey.getBytes());
        BigInteger allowanceBalance;
        if (allowanceBalanceByte == null) {
            allowanceBalance = BigInteger.valueOf(0);
        } else {
            allowanceBalance = new BigInteger(allowanceBalanceByte);
        }

        BigInteger allowanceBalanceNow = allowanceBalance.add(tokenAmount);

        ctx.putObject(allowanceKey.getBytes(), allowanceBalanceNow.toByteArray());

        return Response.ok("approve success".getBytes());
    }

    public static void main(String[] args) {
        Driver.serve(new Erc20());
    }
}
